### PR TITLE
Gas price

### LINKED ISSUE #10: Implement recommended gas price
https://docs.google.com/presentation/d/1hCt72FKBthEtCTWRQvjfvjctbJz3MLHIooE7rP20eZA/edit?pli=1#slide=id.g282ed5981d5_0_117

TODO: Update the issue description to what exactly needs to be done (the above is a private presentation).

### PR DESCRIPTION
**Pull request by vponline**
Closes #10

### CODE REVIEW

File: src/gas-price/gas-price.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    gasPriceStore[chainId]!.lastUpdateTimestamp < Date.now() - scalingWindow * 60 * 1_000
+  ) {
+    return multiplyGasPrice(gasPrice, scalingMultiplier);
+  }
+
+  // Check if the multiplied gas price is within the percentile and return the smaller value
+  // TODO should we check for a minimum length of state gas prices used in the calculation?
+  const sanitizedGasPrice = percentileGasPrice && gasPrice.gt(percentileGasPrice) ? percentileGasPrice : gasPrice;
+
+  // const multipliedGasPrice = multiplyGasPrice(sanitizedGasPrice, recommendedGasPriceMultiplier);
```
Comment (aquarat) [inline src/gas-price/gas-price.ts:150]: Is this related to the comment above?

File: src/gas-price/gas-price.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  const fetchInterval = /*config.fetchInterval*/ 30 * 1_000;
+
+  if (!gasPriceCollectorInterval) {
+    gasPriceCollectorInterval = setInterval(runGasPriceCollector, fetchInterval);
+  }
+
+  await Promise.all(
+    Object.entries(config.chains).map(([chainId, chain]) =>
+      go(
+        async () =>
```
Comment (aquarat) [inline src/gas-price/gas-price.ts:167]: AFAIK this is no longer necessary because the new go utils returns the error correctly, might be worth double-checking that.

File: src/gas-price/gas-price.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+          await airseekerV2ProviderRecommendedGasPrice(
+            chainId,
+            // TODO: what to do with many providers?
```
Comment (aquarat) [inline src/gas-price/gas-price.ts:170]: Implement our our Provider that can do agregation 😆

**Review by aquarat — APPROVED** [review]
LGTM :)

**Comment by vponline** [comment]
Please re-review based on the new spec 😄 

The changes include:

- No separate interval for the gas collector, it will be ran within the main loop
- Separate logic to fetch new price and clear old prices
- Separate logic to fetch new price and calculate airseekerV2 price (with scaling and sanitization)
- Checking for pending transactions based on datafeedId:value mapping
- Added function to clear datafeedId:value mapping

File: src/gas-price/gas-price.ts [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+ * @param providerName
+ * @param sanitizationSamplingWindow
+ */
+export const clearExpiredStoreGasPrices = (
+  chainId: string,
+  providerName: string,
+  sanitizationSamplingWindow: number
+) => {
+  // Remove gasPrices older than the sanitizationSamplingWindow
+  gasPriceStore[chainId]![providerName]!.gasPrices = gasPriceStore[chainId]![providerName]!.gasPrices.filter(
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:53]: I wonder how efficient this will need to be. For now we can leave it as is, but it may be a good idea to put these records in a data structure that supports faster removals.

File: src/gas-price/gas-price.ts [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+  );
+};
+
+/**
+ * Saves last transaction details into the store.
+ * @param chainId
+ * @param providerName
+ * @param nonce
+ */
+export const setLastOnChainDatafeedValues = (
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:64]: Why do we need this? I do not see how data feed values are related to the gas price oracle.

File: src/gas-price/gas-price.ts [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  if (!gasPriceStore[chainId]![providerName]) {
+    gasPriceStore[chainId]![providerName] = { gasPrices: [], lastOnChainDataFeedValues: {} };
+  }
+};
+
+/**
+ * Saves a gas price into the store.
+ * @param chainId
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:30]: nit: I am not a fan of JS Doc comments in typescript, since you can see the names and types of the arguments by default. I like the small text description though.

File: src/gas-price/gas-price.ts [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+/**
+ *  Calculates the gas price to be used in a transaction based on sanitization and scaling settings.
+ * @param chainId
+ * @param providerName
+ * @param rpcUrl
+ * @param gasSettings
+ * @param nonce
+ * @returns {ethers.BigNumber}
+ */
+export const airseekerV2ProviderRecommendedGasPrice = async (
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:167]: I'd just call this `getAirseekerRecommendedGasPrice`, but I understand why you called it that way.

**Review by Siegrift — COMMENTED** [review]
The gas oracle implementation looks good to me, but I think the integration with the main loop will make some functions unused (but we can remove them then).

File: src/gas-price/gas-price.ts [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+
+/**
+ * Saves a gas price into the store.
+ * @param chainId
+ * @param providerName
+ * @param gasPrice
+ */
+export const setStoreGasPrices = (chainId: string, providerName: string, gasPrice: ethers.BigNumber) => {
+  gasPriceStore[chainId]![providerName]!.gasPrices = [
+    { price: gasPrice, timestamp: Date.now() },
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:36]: When we say `timestamp` we usually refer to UNIX timestamp (in seconds). I'd rename this one to `timestampMs` just to be explicit or convert to seconds.

File: src/gas-price/gas-price.ts [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+ * @returns {ethers.BigNumber}
+ */
+export const updateGasPriceStore = async (chainId: string, providerName: string, rpcUrl: string) => {
+  const provider = new ethers.providers.StaticJsonRpcProvider(rpcUrl, {
+    chainId: Number.parseInt(chainId, 10),
+    name: chainId,
+  });
+
+  // Get the provider recommended gas price
+  const goGasPrice = await go(async () => provider.getGasPrice(), { retries: 1, attemptTimeoutMs: 2000 });
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:127]: Not sure how I feel about retrying the get gas price call. It seems a bit arbitrary to me.

File: src/gas-price/gas-price.ts [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+): Promise<ethers.BigNumber> => {
+  const { recommendedGasPriceMultiplier, sanitizationPercentile, scalingWindow, scalingMultiplier } = gasSettings;
+  // Get the configured percentile of historical gas prices before adding the new price
+  const percentileGasPrice = getPercentile(
+    sanitizationPercentile,
+    gasPriceStore[chainId]![providerName]!.gasPrices.map((gasPrice) => gasPrice.price)
+  );
+
+  const gasPrice = await updateGasPriceStore(chainId, providerName, rpcUrl);
+  if (!gasPriceStore[chainId]![providerName]!.gasPrices) {
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:185]: This doesn't make sense to me. This only runs if we don't have any `gasPrices` which happens when we couldn't fetch any gas price from a provider. If that happens the `updateGasPriceStore` throws.

Related to the above, the error handling should be clearer (which function is responsible for handling the error).

File: src/gas-price/gas-price.ts [hunk 11]
```diff
@@ -0,0 +1,236 @@
...
+    return multiplyGasPrice(gasPrice, recommendedGasPriceMultiplier);
+  }
+
+  const lastDataFeedValue =
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:189]: I thought we decided to use nonces to track updates.

File: .github/workflows/main.yml [hunk 12]
```diff
@@ -42,10 +42,31 @@ jobs:
         run: pnpm run prettier:check && pnpm run eslint:check
       - name: Test
         run: pnpm run test
+  
+  test-e2e:
+    name: Test e2e
+    runs-on: ubuntu-latest
+    # Don't run twice for a push within an internal PR
```
[the remark this answers is not in the document]
Comment (vponline) [reply .github/workflows/main.yml:49]: It was included in `operations-database` to reduce the GA runtime since we hit the limit a few times. That's probably still the case?

File: src/gas-price/gas-price.ts [hunk 13]
```diff
@@ -0,0 +1,236 @@
...
+  }
+
+  // Check if the multiplied gas price is within the percentile and return the smaller value
+  // TODO should we check for a minimum length of state gas prices used in the calculation?
+  const sanitizedGasPrice = percentileGasPrice && gasPrice.gt(percentileGasPrice) ? percentileGasPrice : gasPrice;
+
+  return multiplyGasPrice(sanitizedGasPrice, recommendedGasPriceMultiplier);
+};
+
+export const runGasPriceCollector = async () => {
```
[the remark this answers is not in the document]
Comment (vponline) [reply src/gas-price/gas-price.ts:216]: True! Will remove the loop entirely 😄

File: test/e2e/gas-price.feature.ts [hunk 14]
```diff
@@ -0,0 +1,147 @@
...
+  await wallet.sendTransaction({
+    to: hre.ethers.constants.AddressZero,
+    ...(gasPriceOverride ? { gasPrice: gasPriceOverride } : {}),
+  });
+};
+
+describe('airseekerV2ProviderRecommendedGasPrice', () => {
+  beforeEach(async () => {
+    // Reset the local hardhat network state for each test to prevent issues with other test contracts
+    await hre.network.provider.send('hardhat_reset');
```
[the remark this answers is not in the document]
Comment (vponline) [reply test/e2e/gas-price.feature.ts:39]: It's useful to reset the provider recommended gas price since otherwise it would fluctuate between tests. We also have the [runInBand](https://jestjs.io/docs/cli#--runinband) option set to run them sequentially.

File: .github/workflows/main.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (Siegrift) [reply .github/workflows/main.yml:49]: Yeah, why not apply it to all tests then?

File: test/e2e/gas-price.feature.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (Siegrift) [reply test/e2e/gas-price.feature.ts:39]: Thanks. I forgot we added that in the other PR.

File: src/gas-price/gas-price.ts (hunk 5 above)
Comment (vponline) [reply src/gas-price/gas-price.ts:64]: This is used to set the datafeedId:datafeedValue mapping to check for pending transactions (based on the value and timestamp)

File: src/gas-price/gas-price.ts (hunk 11 above)
Comment (vponline) [reply src/gas-price/gas-price.ts:189]: It's not guaranteed to work if we use an Airseeker with a single sponsorWallet for multiple datafeeds (and Burak requested that it would work for the v1 way in [this thread](https://api3workspace.slack.com/archives/C05TQPT7PNJ/p1697721456428479?thread_ts=1697707242.099579&cid=C05TQPT7PNJ))

File: src/gas-price/gas-price.ts (hunk 9 above)
Comment (vponline) [reply src/gas-price/gas-price.ts:129]: Errors should probably be handled at the `getAirseekerV2ProviderRecommendedGasPrice` level in the main loop, with no `go` around the `getGasPrice` if we don't want to retry there.

File: src/gas-price/gas-price.ts [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+  timestampMs: number;
+}
+interface GasState {
+  gasPrices: { price: ethers.BigNumber; timestampMs: number }[];
+  lastOnChainDataFeedValues: Record<string, DataFeedValue>;
+}
+
+export const gasPriceStore: Record<string, Record<string, GasState>> = {};
+
+export const initializeGasStore = (chainId: string, providerName: string) => {
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:16]: I think we should have just a single state, so I'd move merge this logic with `src/state`.

File: src/gas-price/gas-price.ts [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+  const gasPrice = await updateGasPriceStore(chainId, providerName, rpcUrl);
+
+  const lastDataFeedValue =
+    newDataFeedUpdateOnChainValues &&
+    gasPriceStore[chainId]![providerName]!.lastOnChainDataFeedValues[newDataFeedUpdateOnChainValues.dataFeedId];
+  // Check if the next update is a retry of a pending transaction and if it has been pending longer than scalingWindow
+  if (
+    lastDataFeedValue &&
+    newDataFeedUpdateOnChainValues &&
+    lastDataFeedValue?.value === newDataFeedUpdateOnChainValues.newDataFeedValue.value &&
```
Comment (Siegrift) [inline src/gas-price/gas-price.ts:187]: In this [slack comment](https://api3workspace.slack.com/archives/C05TQPT7PNJ/p1698222551839029?thread_ts=1698220736.112239&cid=C05TQPT7PNJ) I keep wondering about this. 

We will be making new transactions all the time (since the singed data changes asynchronously) and this will not track pending txs correctly. I do not see any other way except relying on wallet nonces or awaiting the transaction promise.

File: src/gas-price/gas-price.ts (hunk 15 above)
Comment (vponline) [reply src/gas-price/gas-price.ts:6]: Done, that was quite a refactor 😄

File: test/e2e/gas-price.feature.ts
Comment (Siegrift) [inline test/e2e/gas-price.feature.ts:1]: I really like these test btw.

File: src/gas-price/gas-price.ts (hunk 16 above)
Comment (Siegrift) [reply src/gas-price/gas-price.ts:261]: I guess the important part is that if the on-chain value changes there had to be an update. If that's only what we care about then the logic is sound.

**Review by Siegrift — APPROVED** [review]
🙏 Thanks for the discussions and making this work.

With https://github.com/api3dao/airseeker-v2/pull/26/files#r1372062174 I think the logic is now clear to me.
